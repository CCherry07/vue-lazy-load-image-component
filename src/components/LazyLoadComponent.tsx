import { reactive, computed, watch, defineComponent, ExtractPropTypes } from 'vue'
import PlaceholderWithoutTracking from './PlaceholderWithoutTracking.tsx'
import isIntersectionObserverAvailable from '../utils/intersection-observer';
import PlaceholderWithTracking from './PlaceholderWithTracking.tsx'
import { LazyLoadComponentPropsFunc } from "./interface.ts";

export type LazyLoadComponentProps = Partial<ExtractPropTypes<ReturnType<typeof LazyLoadComponentPropsFunc>>>

export default defineComponent({
  compatConfig: { MODE: 3 },
  name: 'LazyLoadComponent',
  inheritAttrs: false,
  props: LazyLoadComponentPropsFunc(),
  setup(props, { slots }) {
    const state = reactive({
      visible: props.visibleByDefault ?? false,
      loaded: false,
      error: false,
    })
    const isScrollTracked = computed(() => Boolean(
      props.scrollPosition &&
      Number.isFinite(props.scrollPosition.x) &&
      props.scrollPosition.x >= 0 &&
      Number.isFinite(props.scrollPosition.y) &&
      props.scrollPosition.y >= 0
    ))

    watch(state,
      (prevState, state) => {
        if (prevState.visible !== state?.visible) {
          props.afterLoad?.();
        }
      },
      { immediate: true }
    )

    const onVisible = () => {
      state.visible = true
      props.afterLoad?.()
    }

    if (props.visibleByDefault) {
      props.beforeLoad?.()
      props.afterLoad?.()
    }
    return () => {
      if (state.visible) {
        return slots.default?.()
      }
      return isScrollTracked ||
        (props.useIntersectionObserver && isIntersectionObserverAvailable()) ?
        <PlaceholderWithoutTracking
          height={props.height}
          onVisible={onVisible}
          placeholder={props.placeholder}
          scrollPosition={props.scrollPosition}
          threshold={props.threshold}
          useIntersectionObserver={props.useIntersectionObserver}
          width={props.width}
        />
        :
        <PlaceholderWithTracking
          height={props.height}
          onVisible={onVisible}
          placeholder={props.placeholder}
          scrollPosition={props.scrollPosition}
          threshold={props.threshold}
          useIntersectionObserver={props.useIntersectionObserver}
          width={props.width}
        />
    }
  }
})

