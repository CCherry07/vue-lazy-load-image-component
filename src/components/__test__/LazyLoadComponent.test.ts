// LazyLoadComponent test
import { mount } from '@vue/test-utils';
import { expect } from 'vitest';
import LazyLoadComponent from '../LazyLoadComponent.tsx';

test('LazyLoadComponent', async () => {
  const wrapper = mount(LazyLoadComponent, {
    props: {
      height: 100,
      width: 100,
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});
