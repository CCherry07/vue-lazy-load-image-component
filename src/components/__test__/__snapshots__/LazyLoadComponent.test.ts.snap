// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LazyLoadComponent 1`] = `"<span style=\\"display: inline-block; height: 100px; width: 100px;\\"><!----></span>"`;
